File: lua/antybrowser.lua
local json = require("cjson")
local http = require("socket.http")
local ltn12 = require("ltn12")

local M = {}
M.__index = M

function M.new(api_key, opts)
    opts = opts or {}
    local self = setmetatable({}, M)
    self.api_key = api_key
    self.base_url = opts.base_url or ("http://127.0.0.1:" .. (opts.port or 5173))
    return self
end

local function request(self, method, path, body)
    local url = self.base_url .. path
    local resp_body = {}
    local req = {
        url = url,
        method = method,
        headers = {
            ["x-api-key"] = self.api_key,
            ["Content-Type"] = "application/json",
        },
        sink = ltn12.sink.table(resp_body),
    }
    if body then
        req.source = ltn12.source.string(json.encode(body))
    end
    local res, code, headers, status = http.request(req)
    local raw = table.concat(resp_body)
    if code < 200 or code >= 300 then
        return nil, "HTTP " .. code .. ": " .. raw
    end
    return json.decode(raw)
end

-- System
function M:get_status() return request(self, "GET", "/api/status") end
function M:get_settings() return request(self, "GET", "/api/settings") end
function M:get_sync_status() return request(self, "GET", "/api/sync/status") end

function M:refresh_sync(profile_id)
    local body = profile_id and { profileId = profile_id } or {}
    return request(self, "POST", "/api/sync/refresh", body)
end

-- Profiles
function M:get_profiles() return request(self, "GET", "/api/profiles") end
function M:create_profile(data) return request(self, "POST", "/api/profiles", data) end
function M:update_profile(id, data) return request(self, "PUT", "/api/profiles/" .. id, data) end
function M:delete_profile(id) return request(self, "DELETE", "/api/profiles/" .. id) end
function M:start_profile(id) return request(self, "POST", "/api/profiles/" .. id .. "/start") end
function M:stop_profile(id) return request(self, "POST", "/api/profiles/" .. id .. "/stop") end
function M:duplicate_profile(id, name)
    return request(self, "POST", "/api/profiles/" .. id .. "/duplicate", name and { name = name } or nil)
end

-- Automations
function M:get_automations() return request(self, "GET", "/api/automations") end
function M:run_automation(id, profile_id)
    return request(self, "POST", "/api/automations/" .. id .. "/run", { profileId = profile_id })
end

-- Groups
function M:get_groups() return request(self, "GET", "/api/groups") end
function M:create_group(data) return request(self, "POST", "/api/groups", data) end
function M:update_group(id, data) return request(self, "PUT", "/api/groups/" .. id, data) end
function M:delete_group(id) return request(self, "DELETE", "/api/groups/" .. id) end

-- Proxies
function M:get_proxies() return request(self, "GET", "/api/proxies") end
function M:create_proxy(data) return request(self, "POST", "/api/proxies", data) end
function M:check_proxy(data) return request(self, "POST", "/api/proxies/check", data) end
function M:delete_proxy(id) return request(self, "DELETE", "/api/proxies/" .. id) end

-- Extensions
function M:get_extensions() return request(self, "GET", "/api/extensions") end
function M:delete_extension(id) return request(self, "DELETE", "/api/extensions/" .. id) end
function M:get_profile_extensions(profile_id) return request(self, "GET", "/api/profiles/" .. profile_id .. "/extensions") end
function M:set_profile_extensions(profile_id, extension_ids)
    return request(self, "POST", "/api/profiles/" .. profile_id .. "/extensions", { extensionIds = extension_ids })
end

return M

File: elixir/lib/antybrowser.ex
defmodule Antybrowser do
  @moduledoc "Official Antybrowser SDK — Elixir client for the Antybrowser Local API"

  defstruct [:api_key, :base_url]

  def new(api_key, opts \\ []) do
    port = Keyword.get(opts, :port, 5173)
    %__MODULE__{
      api_key: api_key,
      base_url: "http://127.0.0.1:#{port}"
    }
  end

  # System
  def get_status(client), do: get(client, "/api/status")
  def get_settings(client), do: get(client, "/api/settings")
  def get_sync_status(client), do: get(client, "/api/sync/status")

  def refresh_sync(client, profile_id \\ nil) do
    body = if profile_id, do: %{profileId: profile_id}, else: %{}
    post(client, "/api/sync/refresh", body)
  end

  # Profiles
  def get_profiles(client), do: get(client, "/api/profiles")
  def create_profile(client, data), do: post(client, "/api/profiles", data)
  def update_profile(client, id, data), do: put(client, "/api/profiles/#{id}", data)
  def delete_profile(client, id), do: delete(client, "/api/profiles/#{id}")
  def start_profile(client, id), do: post(client, "/api/profiles/#{id}/start", %{})
  def stop_profile(client, id), do: post(client, "/api/profiles/#{id}/stop", %{})

  def duplicate_profile(client, id, name \\ nil) do
    body = if name, do: %{name: name}, else: %{}
    post(client, "/api/profiles/#{id}/duplicate", body)
  end

  # Automations
  def get_automations(client), do: get(client, "/api/automations")
  def run_automation(client, id, profile_id), do: post(client, "/api/automations/#{id}/run", %{profileId: profile_id})

  # Groups
  def get_groups(client), do: get(client, "/api/groups")
  def create_group(client, data), do: post(client, "/api/groups", data)
  def update_group(client, id, data), do: put(client, "/api/groups/#{id}", data)
  def delete_group(client, id), do: delete(client, "/api/groups/#{id}")

  # Proxies
  def get_proxies(client), do: get(client, "/api/proxies")
  def create_proxy(client, data), do: post(client, "/api/proxies", data)
  def check_proxy(client, data), do: post(client, "/api/proxies/check", data)
  def delete_proxy(client, id), do: delete(client, "/api/proxies/#{id}")

  # Extensions
  def get_extensions(client), do: get(client, "/api/extensions")
  def delete_extension(client, id), do: delete(client, "/api/extensions/#{id}")
  def get_profile_extensions(client, profile_id), do: get(client, "/api/profiles/#{profile_id}/extensions")

  def set_profile_extensions(client, profile_id, extension_ids) do
    post(client, "/api/profiles/#{profile_id}/extensions", %{extensionIds: extension_ids})
  end

  # HTTP
  defp get(client, path), do: request(client, :get, path, nil)
  defp post(client, path, body), do: request(client, :post, path, body)
  defp put(client, path, body), do: request(client, :put, path, body)
  defp delete(client, path), do: request(client, :delete, path, nil)

  defp request(client, method, path, body) do
    url = client.base_url <> path
    headers = [{"x-api-key", client.api_key}, {"Content-Type", "application/json"}]

    result = case method do
      :get -> HTTPoison.get(url, headers)
      :post -> HTTPoison.post(url, Jason.encode!(body), headers)
      :put -> HTTPoison.put(url, Jason.encode!(body), headers)
      :delete -> HTTPoison.delete(url, headers)
    end

    case result do
      {:ok, %HTTPoison.Response{status_code: code, body: resp_body}} when code in 200..299 ->
        {:ok, Jason.decode!(resp_body)}
      {:ok, %HTTPoison.Response{status_code: code, body: resp_body}} ->
        {:error, %{status_code: code, body: resp_body}}
      {:error, reason} ->
        {:error, reason}
    end
  end
end
